components/style/stylist.rs

```diff
@@ -98,15 +98,24 @@ impl DocumentCascadeData {
     {
         debug_assert!(!flusher.nothing_to_do());
 
-        let origins_to_rebuild = flusher.origins_to_fully_rebuild();
+        for (cascade_data, origin) in self.per_origin.iter_mut_origins() {
+            let validity = flusher.origin_validity(origin);
 
-        for origin in origins_to_rebuild.iter() {
-            extra_data.borrow_mut_for_origin(&origin).clear();
-            self.per_origin.borrow_mut_for_origin(&origin).clear();
-        }
+            if validity == OriginValidity::Valid {
+                continue;
+            }
+
+            if origin == Origin::UserAgent {
+                self.precomputed_pseudo_element_decls.clear();
+            }
 
-        if origins_to_rebuild.contains(Origin::UserAgent.into()) {
-            self.precomputed_pseudo_element_decls.clear();
+            extra_data.borrow_mut_for_origin(&origin).clear();
+            if validity == OriginValidity::CascadeInvalid {
+                cascade_data.clear_cascade_data()
```

Comment:
Hmm, now these names are slightly confusing I guess.  ("Cascade data" both for the top level, per-origin data, and also for the tables which represent performing the cascading and not the invalidations...)